### PR TITLE
[BugFix] fix timestamp underflow for parquet format

### PR DESCRIPTION
**Pull request by dirtysalt**
## What type of PR is this：
- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto backported to target branch
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3

### CODE REVIEW

File: be/src/formats/parquet/column_converter.cpp [hunk 1]
```diff
@@ -682,9 +682,16 @@ Status Int64ToDateTimeConverter::_convert_to_timestamp_column(const ColumnPtr& s
...
         if (!src_null_data[i]) {
-            Timestamp timestamp =
-                    timestamp::of_epoch_second(static_cast<int>(src_data[i] / _second_mask),
-                                               static_cast<int>((src_data[i] % _second_mask) * _scale_to_nano_factor));
+            int64_t seconds = src_data[i] / _second_mask;
+            // for the case if seconds < 0(anytime before 1970-01-01 00:00:00)
+            // like '1900-01-01 00:00:00', we will get '1899-12-31 23:54:17'
+            // there is 343 seconds behind somehow.(But I really don't know why)
+            // so to get correct answer, we have to add that back.
+            if (seconds < 0) {
```
Comment (imay) [inline be/src/formats/parquet/column_converter.cpp:690]: If the second is -1, the result will be larger than utc? Is that correct?

Comment (dirtysalt) [reply be/src/formats/parquet/column_converter.cpp:690]: if second is -1, you expect to get '1969-12-31 23:59:59', However you probably will get `1969-12-31 23:54:16`(343 seconds are lost). So to offset that, I'll add 343 seconds.

Comment (Seaven) [reply be/src/formats/parquet/column_converter.cpp:690]: emmm.... I think your method `of_epoch_second` is error...the method compute julian date is error
your refs this code:
```
    TimestampValue ep;
    ep.from_unix_second(-1);

    auto js = timestamp::UNIX_EPOCH_SECONDS -1;
    jd = js / SECS_PER_DAY;

    std::cout << "ep: " << ep._timestamp
              << ", str: " << ep.to_string() 
              << ", jd:" << jd << std::endl;
```

**Review by Seaven — CHANGES_REQUESTED** [review]

File: be/src/formats/parquet/column_converter.cpp (hunk 1 above)
Comment (dirtysalt) [reply be/src/formats/parquet/column_converter.cpp:690]: thanks, I'll have a try.

Comment (dirtysalt) [reply be/src/formats/parquet/column_converter.cpp:690]: I've tried following code, but there is still 343 seconds lost.

```cpp

Timestamp timestamp::of_epoch_second(int64_t seconds, int64_t nanoseconds) {
    int64_t second = seconds + timestamp::UNIX_EPOCH_SECONDS;
    JulianDate day = second / SECS_PER_DAY;
    Timestamp s = second % SECS_PER_DAY;
    return timestamp::from_julian_and_time(day, s * USECS_PER_SEC + nanoseconds / NANOSECS_PER_USEC);
}
```

**Comment by dirtysalt** [comment]
@imay @Seaven I've added a ut to reproduce this case.

**Comment by dirtysalt** [comment]
@Seaven recommends to check following dates' timestamp? Maybe becuase of leap second? 

1970-1-1 00:00:00, 1960-1-1 00:00:00,1950-1-1 00:00:00,1940-1-1 00:00:00,1930-1-1 00:00:00,1920-1-1 00:00:00

**Comment by dirtysalt** [comment]
run starrocks_admit_test

**Comment by dirtysalt** [comment]
create table 
```

CREATE TABLE `test_hudi_mor16` (
  `uuid` STRING,
  `ts` int,
  `a` int,
  `b` timestamp)
  USING hudi
TBLPROPERTIES (
  'primaryKey' = 'uuid',
  'preCombineField' = 'ts',
  'type' = 'mor');

```

write data using spark

```

import datetime
conf = spark.conf
conf.set(    'spark.sql.avro.datetimeRebaseModeInWrite', 'CORRECTED')
conf.set(    'spark.sql.parquet.datetimeRebaseModeInWrite', 'CORRECTED')
inserts = []
for i in range(10):
    inserts.append({'uuid': 'aa%d' % i, 'ts': 20 + i, 'a': 100 + i, 'b': datetime.datetime(1900+10 * i, 1,1,0,0,0)})
df = spark.createDataFrame(sc.parallelize(inserts, 2))
basePath = "hdfs://emr-header-1.cluster-49091:9000/user/hive/warehouse/zya.db/test_hudi_mor16/";
tableName = 'zya.test_hudi_mor16';
hudi_options = {
    'hoodie.table.name': tableName,
    'hoodie.datasource.write.recordkey.field': 'uuid',
    'hoodie.datasource.write.table.name': tableName,
    'hoodie.datasource.write.operation': 'upsert',
    'hoodie.datasource.write.precombine.field': 'ts',
    'hoodie.upsert.shuffle.parallelism': 2,
    'hoodie.insert.shuffle.parallelism': 2,
    'spark.sql.avro.datetimeRebaseModeInWrite': 'CORRECTED',
    'spark.sql.parquet.datetimeRebaseModeInWrite': 'CORRECTED'
}

df.write.format("hudi"). \
    options(**hudi_options). \
    mode("overwrite"). \
    save(basePath)

```

select content

```
spark-sql> select * from test_hudi_mor16;
_hoodie_commit_time     _hoodie_commit_seqno    _hoodie_record_key      _hoodie_partition_path  _hoodie_file_name       a       b       ts      uuid
20230301063631309       20230301063631309_0_0   aa7             8162dac1-6ff7-47f8-8253-75075d18611e-0_0-30-31_20230301063631309.parquet        107     1970-01-01 00:00:00     27      aa7
20230301063631309       20230301063631309_0_1   aa1             8162dac1-6ff7-47f8-8253-75075d18611e-0_0-30-31_20230301063631309.parquet        101     1910-01-01 00:00:00     21      aa1
20230301063631309       20230301063631309_0_2   aa3             8162dac1-6ff7-47f8-8253-75075d18611e-0_0-30-31_20230301063631309.parquet        103     1930-01-01 00:00:00     23      aa3
20230301063631309       20230301063631309_0_3   aa5             8162dac1-6ff7-47f8-8253-75075d18611e-0_0-30-31_20230301063631309.parquet        105     1950-01-01 00:00:00     25      aa5
20230301063631309       20230301063631309_0_4   aa9             8162dac1-6ff7-47f8-8253-75075d18611e-0_0-30-31_20230301063631309.parquet        109     1990-01-01 00:00:00     29      aa9
20230301063631309       20230301063631309_0_5   aa4             8162dac1-6ff7-47f8-8253-75075d18611e-0_0-30-31_20230301063631309.parquet        104     1940-01-01 00:00:00     24      aa4
20230301063631309       20230301063631309_0_6   aa8             8162dac1-6ff7-47f8-8253-75075d18611e-0_0-30-31_20230301063631309.parquet        108     1980-01-01 00:00:00     28      aa8
20230301063631309       20230301063631309_0_7   aa0             8162dac1-6ff7-47f8-8253-75075d18611e-0_0-30-31_20230301063631309.parquet        100     1900-01-01 00:00:00     20      aa0
20230301063631309       20230301063631309_0_8   aa6             8162dac1-6ff7-47f8-8253-75075d18611e-0_0-30-31_20230301063631309.parquet        106     1960-01-01 00:00:00     26      aa6
20230301063631309       20230301063631309_0_9   aa2             8162dac1-6ff7-47f8-8253-75075d18611e-0_0-30-31_20230301063631309.parquet        102     1920-01-01 00:00:00     22      aa2
```

parquet content

```
_hoodie_commit_time = 20230301063631309
_hoodie_commit_seqno = 20230301063631309_0_0
_hoodie_record_key = aa7
_hoodie_partition_path = 
_hoodie_file_name = 8162dac1-6ff7-47f8-8253-75075d18611e-0_0-30-31_20230301063631309.parquet
a = 107
b = -28800000000
ts = 27
uuid = aa7

_hoodie_commit_time = 20230301063631309
_hoodie_commit_seqno = 20230301063631309_0_1
_hoodie_record_key = aa1
_hoodie_partition_path = 
_hoodie_file_name = 8162dac1-6ff7-47f8-8253-75075d18611e-0_0-30-31_20230301063631309.parquet
a = 101
b = -1893484800000000
ts = 21
uuid = aa1

_hoodie_commit_time = 20230301063631309
_hoodie_commit_seqno = 20230301063631309_0_2
_hoodie_record_key = aa3
_hoodie_partition_path = 
_hoodie_file_name = 8162dac1-6ff7-47f8-8253-75075d18611e-0_0-30-31_20230301063631309.parquet
a = 103
b = -1262332800000000
ts = 23
uuid = aa3

_hoodie_commit_time = 20230301063631309
_hoodie_commit_seqno = 20230301063631309_0_3
_hoodie_record_key = aa5
_hoodie_partition_path = 
_hoodie_file_name = 8162dac1-6ff7-47f8-8253-75075d18611e-0_0-30-31_20230301063631309.parquet
a = 105
b = -631180800000000
ts = 25
uuid = aa5

_hoodie_commit_time = 20230301063631309
_hoodie_commit_seqno = 20230301063631309_0_4
_hoodie_record_key = aa9
_hoodie_partition_path = 
_hoodie_file_name = 8162dac1-6ff7-47f8-8253-75075d18611e-0_0-30-31_20230301063631309.parquet
a = 109
b = 631123200000000
ts = 29
uuid = aa9

_hoodie_commit_time = 20230301063631309
_hoodie_commit_seqno = 20230301063631309_0_5
_hoodie_record_key = aa4
_hoodie_partition_path = 
_hoodie_file_name = 8162dac1-6ff7-47f8-8253-75075d18611e-0_0-30-31_20230301063631309.parquet
a = 104
b = -946800000000000
ts = 24
uuid = aa4

_hoodie_commit_time = 20230301063631309
_hoodie_commit_seqno = 20230301063631309_0_6
_hoodie_record_key = aa8
_hoodie_partition_path = 
_hoodie_file_name = 8162dac1-6ff7-47f8-8253-75075d18611e-0_0-30-31_20230301063631309.parquet
a = 108
b = 315504000000000
ts = 28
uuid = aa8

_hoodie_commit_time = 20230301063631309
_hoodie_commit_seqno = 20230301063631309_0_7
_hoodie_record_key = aa0
_hoodie_partition_path = 
_hoodie_file_name = 8162dac1-6ff7-47f8-8253-75075d18611e-0_0-30-31_20230301063631309.parquet
a = 100
b = -2209017943000000
ts = 20
uuid = aa0

_hoodie_commit_time = 20230301063631309
_hoodie_commit_seqno = 20230301063631309_0_8
_hoodie_record_key = aa6
_hoodie_partition_path = 
_hoodie_file_name = 8162dac1-6ff7-47f8-8253-75075d18611e-0_0-30-31_20230301063631309.parquet
a = 106
b = -315648000000000
ts = 26
uuid = aa6

_hoodie_commit_time = 20230301063631309
_hoodie_commit_seqno = 20230301063631309_0_9
_hoodie_record_key = aa2
_hoodie_partition_path = 
_hoodie_file_name = 8162dac1-6ff7-47f8-8253-75075d18611e-0_0-30-31_20230301063631309.parquet
a = 102
b = -1577952000000000
ts = 22
uuid = aa2

```

**Comment by dirtysalt** [comment]
1900-01-01 00:00:00 --->  -2209017943000000
1910-01-01 00:00:00 --->  -1893484800000000
1920-01-01 00:00:00 --->  -1577952000000000
1930-01-01 00:00:00 --->  -1262332800000000
1940-01-01 00:00:00 --->  -946800000000000
1950-01-01 00:00:00 --->  -631180800000000
1960-01-01 00:00:00 --->  -315648000000000
1970-01-01 00:00:00 --->  -28800000000
1980-01-01 00:00:00 --->  315504000000000
1990-01-01 00:00:00 --->  631123200000000

**Comment by Seaven** [comment]
I think the answer in here: 
https://stackoverflow.com/questions/67568883/why-does-the-time-package-have-different-behaviors-when-date-is-before-1900/67569074#67569074

**Comment by Seaven** [comment]
so, time convert need consider timezone, can't use unixtime direct

**Review by Seaven — DISMISSED** [review]

**Comment by dirtysalt** [comment]
run starrocks_be_unittest

**Review by stephen-shelby — DISMISSED** [review]

**Review by Youngwb — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by dirtysalt** [comment]
run starrocks_be_unittest

**Comment by dirtysalt** [comment]
run starrocks_admit_test

**Comment by dirtysalt** [comment]
run starrocks_be_unittest

**Comment by dirtysalt** [comment]
run starrocks_admit_test

**Review by satanson — APPROVED** [review]

**Review by wyb — APPROVED** [review]

**Comment by dirtysalt** [comment]
@MergifyIO backport branch-2.5

**Comment by dirtysalt** [comment]
@Mergify backport branch-3.0

**Comment by eshishki** [comment]
looks like it introduced performance regression

**Pull request merged**